addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java

```diff
@@ -37,89 +36,50 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-        }
-    }
-
-    @Override
     public void run() {
-        final int dataBufferMaxLen = Byte.MAX_VALUE;
-
-        byte[] dataBuffer = new byte[dataBufferMaxLen];
-
-        int msgLen = 0;
-        int index = 0;
-        boolean startFound = false;
-
         logger.debug("Data listener started");
 
-        try {
-
-            byte[] tmpData = new byte[20];
-            int len = -1;
-
-            while (!interrupted) {
-
-                if ((len = in.read(tmpData)) > 0) {
-
-                    byte[] logData = Arrays.copyOf(tmpData, len);
-                    logger.trace("Received data (len={}): {}", len, DatatypeConverter.printHexBinary(logData));
-
-                    for (int i = 0; i < len; i++) {
-
-                        if (index > dataBufferMaxLen) {
-                            // too many bytes received, try to find new
-                            // start
-                            startFound = false;
-                        }
-
-                        if (!startFound && tmpData[i] > 0) {
-
-                            startFound = true;
-                            index = 0;
-                            dataBuffer[index++] = tmpData[i];
-                            msgLen = tmpData[i] + 1;
-
-                        } else if (startFound) {
-
-                            dataBuffer[index++] = tmpData[i];
-
-                            if (index == msgLen) {
-                                sendMessage(dataBuffer, msgLen);
-
-                                // find new start
-                                startFound = false;
-                            }
+        final int MAX_READ_TIMEOUTS = 4;
+        byte[] buf = new byte[Byte.MAX_VALUE];
+
+        // The stream has (or SHOULD have) a read timeout set. Taking a
+        // read timeout (read returns 0) between packets gives us a chance
+        // to check if we've  been interrupted. Read interrupts during a
+        // packet are ignored but if too many occur we take it as meaning
+        // the RFXCOM has become missing presumed dead.
+        while (!Thread.interrupted()) {
+            try {
+                // First byte tells us how long the packet is
+                int bytesRead = in.read(buf, 0, 1);
+                int packetLength = buf[0];
+
+                if (bytesRead > 0 && packetLength > 0) {
+                    // Now read the rest of the packet
+                    int bufferIndex = 1;
+                    int readTimeoutCount = 1;
+
+                    while (bufferIndex <= packetLength) {
+                        int bytesRemaining = packetLength - bufferIndex + 1;
+                        bytesRead = in.read(buf, bufferIndex, bytesRemaining);
+
+                        if (bytesRead > 0) {
+                            bufferIndex += bytesRead;
+                        } else if (readTimeoutCount++ == MAX_READ_TIMEOUTS) {
+                            throw new IOException("Timeout during packet read");
```

Comment:
A continue would be fatal. The inner packet-reading loop (deliberately) does not check for interrupts and read is NOT an interruptible (which was a major problem with the original). If you continue that loop then a stalled (as opposed to I/O erroring) rfxcom becomes non-recoverable. The throw jumps over the handling for the completed packet and into the error handler for the outer loop. This sends the error upstream and continues in the outer loop waiting for a new packet length header. This inter-packet loop DOES check for interrupts and handles shutdown gracefully.

Now sending the error upstream feeds errorOccurred in RFXComBridgeHandler, which logs it and sets the bridge offline. If a packet is then received successfully (remember the stream reader is waiting in hope) then the bridge handler's MessageListener.packetReceived will set it back online. Alternatively the bridge handler has a "check" job that fires every minute. If that job fires before a packet is received it sees the bridge is offline and calls connect(). Connect() disconnects any existing connector (part of which involves interrupting the stream read to tell it to go away) and attempts to connect a new one. If successful we're back online again. Otherwise we just keep trying every minute. Similarly taking an error during sending a message takes the bridge offline and relies on the same packet-received-or-check-job to fix things.